processing/src/main/java/org/apache/carbondata/processing/loading/TableProcessingOperations.java

```diff
@@ -54,39 +59,77 @@
+        String partitionPath = CarbonTablePath.getPartitionDir(carbonTable.getTablePath());
+        if (FileFactory.isFileExist(partitionPath)) {
+          final LoadMetadataDetails[] details =
+              SegmentStatusManager.readLoadMetadata(metaDataLocation);
+          CarbonFile carbonFile = FileFactory.getCarbonFile(partitionPath);
+          CarbonFile[] listFiles = carbonFile.listFiles(new CarbonFileFilter() {
+            @Override public boolean accept(CarbonFile path) {
+              String segmentId = CarbonTablePath.DataFileUtil
+                  .getSegmentIdFromPath(path.getAbsolutePath() + "/dummy");
+              boolean found = false;
+              for (int j = 0; j < details.length; j++) {
+                if (details[j].getLoadName().equals(segmentId)) {
+                  found = true;
+                  break;
+                }
+              }
+              return !found;
+            }
+          });
+          CarbonFile[] allSegmentFiles = null;
+          if (isCompactionFlow) {
+            String segmentFilesLocation =
+                CarbonTablePath.getSegmentFilesLocation(carbonTable.getTablePath());
+            allSegmentFiles = FileFactory.getCarbonFile(segmentFilesLocation).listFiles();
+          }
+          for (int k = 0; k < listFiles.length; k++) {
+            String segmentId = CarbonTablePath.DataFileUtil
+                .getSegmentIdFromPath(listFiles[k].getAbsolutePath() + "/dummy");
+            if (isCompactionFlow) {
+              if (segmentId.contains(".")) {
+                CarbonLoaderUtil.deleteStorePath(listFiles[k].getAbsolutePath());
+                // if segment file exist delete it.
+                for (CarbonFile segment : allSegmentFiles) {
+                  if (segment.getPath().contains(segmentId)) {
+                    segment.delete();
+                    break;
+                  }
+                }
+              }
+            } else {
+              if (!segmentId.contains(".")) {
+                CarbonLoaderUtil.deleteStorePath(listFiles[k].getAbsolutePath());
+              }
             }
           }
-          return !found;
         }
-      });
-      for (int k = 0; k < listFiles.length; k++) {
-        String segmentId = CarbonTablePath.DataFileUtil
-            .getSegmentIdFromPath(listFiles[k].getAbsolutePath() + "/dummy");
+      } else {
         if (isCompactionFlow) {
-          if (segmentId.contains(".")) {
-            CarbonLoaderUtil.deleteStorePath(listFiles[k].getAbsolutePath());
-          }
-        } else {
-          if (!segmentId.contains(".")) {
-            CarbonLoaderUtil.deleteStorePath(listFiles[k].getAbsolutePath());
-          }
+          LOGGER.error(
+              "Not able to acquire the Table status lock for partial load deletion for table "
+                  + carbonTable.getDatabaseName() + "." + carbonTable.getTableName()
+                  + ", retry compaction");
+          throw new RuntimeException(
```

Comment:
Do we really need to throw exception here? because since it's clean up, we can just give an error message and retry next time instead of failing the operation, what do you think?

Reply:
yes, for compaction flow I faced the issue. If the stale segments not cleared. there will be duplicate data after compaction. so, I cannot go ahead for compaction if the stale data is not cleaned. so, throwing the exception here.